### PR TITLE
Prevent `QueryInfo#markResult` mutation of `result.data` and return cache data consistently whether complete or incomplete

### PR DESCRIPTION
**Pull request by benjamn**
This PR fixes issue #9293 by ensuring all cache-friendly fetch policies (including `cache-{first,only,and-network}` and `network-only` but excluding `no-cache`) return result data after re-reading the data from the cache, even when the data are incomplete (fixing #9293), so any custom `read` functions (or other cache logic) can be applied consistently whether or not the cache data are complete. I've added a new test that exercises all the fetch policies, since test coverage was previously inadequate in this area.

While diagnosing this problem, I noticed `QueryInfo#markResult` previously updated `result.data` using a destructive mutation, which could easily go unnoticed by the caller of `markResult`, so I refactored that code a bit to return a new `result` object instead of modifying the one provided.

As @CarsonF's [reproduction](https://github.com/apollographql/apollo-client/issues/9293#issuecomment-1435426921) from #9293 suggests, consistently returning cache data is important when using a `read` function to implement a scalar wrapper like `Date` for an otherwise string-valued field. I tested these changes against the [reproduction repo](https://github.com/CarsonF/apollo-reproduction-read-policy-ignored), and while there's still some weirdness related to the use of `offsetLimitPagination` for the `Query.people` list field, the problem of the disappearing `Date` wrappers seems to be solved. More generally, I believe this change allows a more reliable implementation our most popular feature request (by 👍 reactions), https://github.com/apollographql/apollo-feature-requests/issues/368.

While I stand behind the logic of this change for consistency's sake, it's worth mentioning the old behavior allowed returning raw network results with extraneous fields not present in the query, in cases where the server violates GraphQL validation rules by returning more properties than requested, but also does not return all the requested properties. See my test changes involving `todoWithoutPrice` and `carDataWithoutVine` for examples of this edge case. Now that we always re-read results from the cache using the provided query, you should always receive only the queried fields, so any extraneous properties will (correctly) disappear.

### CODE REVIEW

**Comment by benjamn** [comment]
@jerelmiller @phryneas @alessbell I guess I'm not exactly sure how to update the `size-limit` limits. It looks like `npx size-limit --json` gets run twice, and passes the first time but not the second.

**Comment by benjamn** [comment]
Ran `npm run format` and rebased against @alessbell's `.size-limit.cjs` changes from the base `release-3.9` branch, with a small additional size tweak, and it looks like all tests/checks are passing now, so I think y'all can review this PR now without having to see any red ❌s. Thanks for your help behind the scenes (on Slack).

**Review by phryneas — COMMENTED** [review]
I've adjusted the test a bit and have some questions on this - let's talk about it later!

File: src/core/QueryInfo.ts [hunk 1]
```diff
@@ -408,7 +408,10 @@ export class QueryInfo {
             });
 
             this.lastWrite = {
-              result,
+              // Make a shallow defensive copy of the result object, in case we
+              // later later modify result.data in place, since we don't want
+              // that mutation affecting the saved lastWrite.result.data.
+              result: { ...result },
```
Comment (phryneas) [inline src/core/QueryInfo.ts:414]: Just wanted to add: definitely not challenging this defensive copy - 💯 agree!

File: src/core/__tests__/QueryManager/index.ts [hunk 2]
```diff
@@ -2519,9 +2410,7 @@ describe("QueryManager", () => {
             loading: false,
             networkStatus: NetworkStatus.ready,
             data: {
-              info: {
-                a: "ay",
-              },
+              info: {},
```
[the remark this answers is not in the document]
Comment (benjamn) [reply src/core/__tests__/QueryManager/index.ts:2413]: This is a somewhat weird test that exercises the feud-stopping logic described by the [long comment](https://github.com/apollographql/apollo-client/blob/ae2651a6370c733984cf669099e5c20e87fb05eb/src/core/QueryInfo.ts#L416-L447) in `QueryInfo#markResult`. I ultimately want to remove that logic and this test, but the test currently captures expected behavior.

Specifically, the `data.info.a` property already disappeared in the previous case (`count === 3`) because of writing the `{ info: { b: "bee" }}` result into the cache, clobbering the `info.a` field. Because the most recent `{ info: { a: "ay" }}` result is the same as before, `markResult` skips writing it to the cache (not happy with this compromise, but see the long comment for rationale), so it does not reappear in the cache in the `count === 4` case.

What changed with this PR is that we used to return the raw network result, which does have `{ a: "ay" }`, because the result from the cache would be incomplete. Now, we always return what we get from the cache, which is incomplete (missing the `a: "ay"` property).

If we allowed these two queries to keep clobbering each other's properties (because we can't merge the `Query.info` object, because it has no identity and hasn't been configured with `merge: true`), then we'd get an infinite request loop, which is arguably worse than the weird behavior of this test. Ultimately, I think we should be able to make it easier (even default) for these two queries to have their own separate views of the `Query.info` object, so they don't keep stepping on each other's data, but the standard current solution is to mark the `Query.info` field as mergeable, so `info.a` and `info.b` can coexist:
```ts
new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        info: { merge: true }
      }
    }
  }
})
```

[the remark this answers is not in the document]
Comment (phryneas) [reply src/core/__tests__/QueryManager/index.ts:2413]: Makes sense, thanks for the explanation!

**Review by phryneas — APPROVED** [review]
Approved!

@benjamn we already talked about this - I changed the defensive copy logic in this commit: [`ec5d86c` (#11202)](https://github.com/apollographql/apollo-client/pull/11202/commits/ec5d86c2efb676870da4e1857c1fc578fa32958c)

Could you take another look at that before merging?

File: src/core/QueryInfo.ts [hunk 3]
```diff
@@ -470,20 +476,23 @@ export class QueryInfo {
...
+          // since we have other ways of notifying for this result.
           this.updateLastDiff(diff, diffOptions);
-          if (diff.complete) {
-            result.data = diff.result;
-          }
+          result = {
+            ...result,
+            // TODO Improve types so we don't need this cast.
+            data: diff.result as any,
+          };
```
[the remark this answers is not in the document]
Comment (benjamn) [reply src/core/QueryInfo.ts]: Consistency is good!

**Comment by benjamn** [comment]
Your changes look good to me. Thanks for jumping in and reviewing!
